### PR TITLE
Linear monad transformers

### PR DESCRIPTION
**Pull request by b-mehta**
Adds linear reader and state monad transformers, and a monad transformer class.

### CODE REVIEW

File: src/Control/Monad/Linear.hs [hunk 1]
```diff
@@ -17,10 +17,90 @@ module Control.Monad.Linear
...
+
+
+get :: (Applicative m, Dupable s) => StateT s m s
+get = state dup
+
+put :: (Applicative m, Consumable s) => s ->. StateT s m ()
+put = void . replace
+
+-- XXX: If multiplicity 0 is added, we can have
+-- gets :: Applicative m => (s # 0 -> a) ->. StateT s m a
```
Comment (aspiwack) [inline src/Control/Monad/Linear.hs]: Though that is an incomparable `gets`. So which one is best? It probably shouldn't be in an `XXX` unless we truly have an intention to change it.

File: src/Control/Monad/Linear/Internal.hs [hunk 2]
```diff
@@ -129,5 +132,85 @@ instance Monad m => Monad (NonLinear.ReaderT r m) where
...
+-- | @replace s@ will replace the current state with the new given state, and
+-- return the old state.
+replace :: Applicative m => s ->. StateT s m s
+replace s = state $ (\s' -> (s', s))
+
+-----------------------------
+-- Monad transformer class --
+-----------------------------
+
+class (forall m. Monad m => Monad (t m)) => MonadTrans t where
```
Comment (aspiwack) [inline src/Control/Monad/Linear/Internal.hs]: Yay :slightly_smiling_face: !

File: src/Control/Monad/Linear.hs (hunk 1 above)
Comment (b-mehta) [reply src/Control/Monad/Linear.hs]: Fair enough, I'll remove the comment for now.

**Comment by aspiwack** [comment]
Remember: approved means that you can merge when you see fit. Unless you want a second round of review.

**Pull request merged**